client/json-rpc/src/async_client/client.rs

```diff
@@ -280,53 +376,121 @@ impl<R: RetryStrategy> Client<R> {
         }
     }
 
-    pub async fn send_without_retry(
+    pub async fn send_without_retry(&self, request: &Request) -> Result<JsonRpcResponse, Error> {
+        let rpc_resp: JsonRpcResponse = self.send_json_request(request.to_json(1)).await?;
+        self.validate(rpc_resp)
+    }
+
+    /// Batch requests into one JSON-RPC batch request.
+    /// Batch responses may contain partial success response, or different errors,
+    /// hence this function won't be able to handle errors for caller, including
+    /// retry on specific error.
+    ///
+    /// Caller should handle errors and retry as needed.
+    ///
+    /// After errors are handled, caller can convert `JsonRpcResponse` into `Response<T>`
+    /// in order of requests:
+    ///
+    /// ```rust
+    /// use libra_json_rpc_client::async_client::{
+    ///     types as jsonrpc, Client, Error, Request, Response, Retry,
+    /// };
+    /// use std::convert::{TryInto};
+    ///
+    /// # async fn doc() -> Result<(), Error> {
+    /// let client = Client::from_url("http://testnet.libra.org/v1", Retry::default()).unwrap();
+    /// let mut res = client
+    ///     .batch_send(vec![Request::get_metadata(), Request::get_currencies()])
+    ///     .await?;
+    ///
+    /// let metadata: Response<jsonrpc::Metadata> = res
+    ///     .remove(0)?
+    ///     .try_into()?;
+    ///
+    /// let currencies: Response<jsonrpc::CurrencyInfo> = res
+    ///     .remove(0)?
+    ///     .try_into()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    pub async fn batch_send(
+        &self,
+        requests: Vec<Request>,
+    ) -> Result<Vec<Result<JsonRpcResponse, Error>>, Error> {
+        self.batch_send_without_retry(&requests).await
+    }
+
+    pub async fn batch_send_without_retry(
         &self,
-        method: &str,
-        params: &serde_json::Value,
-    ) -> Result<JsonRpcResponse, Error> {
+        requests: &[Request],
+    ) -> Result<Vec<Result<JsonRpcResponse, Error>>, Error> {
+        let json_requests: Vec<serde_json::Value> = requests
+            .iter()
+            .enumerate()
+            .map(|(i, r)| r.to_json(i))
+            .collect();
+
+        let rpc_resps: Vec<JsonRpcResponse> = self.send_json_request(json![json_requests]).await?;
+
+        let mut result: Vec<Result<JsonRpcResponse, Error>> = json_requests
+            .iter()
+            .map(|r| {
+                Err(Error::unexpected_response_not_found_error(format!(
+                    "request: {}",
+                    r
+                )))
+            })
+            .collect::<Vec<_>>();
+        for resp in rpc_resps {
+            if let Some(ref id) = resp.id {
+                if let Ok(index) = serde_json::from_value::<usize>(id.clone()) {
```

Comment:
don't you want to throw errors if. fail to deserialize?

Reply:
it means response id does not match request id, as request id is index of requests array

Reply:
the problem with reporting error is, we don't know where to put the unmatched response, as batch response may contain valid response. and id mismatched response can't be put into result.